tests.js

```diff
@@ -16,45 +16,167 @@
 
 class Tests {
   constructor(options) {
-    this.items = options.manifest.items
-        .filter((item) => !options.httpOnly || item.protocol === 'http');
-    this.individualItems = options.manifest.individualItems;
+    this.tests = options.tests;
+    this.endpoints = this.buildEndpoints();
     this.host = options.host;
+    this.httpOnly = options.httpOnly;
   }
 
-  list(after, limit) {
-    let begin; let end;
-    if (after) {
-      const afterURL = new URL(after);
-      const afterIndex = this.items.findIndex((item) => {
-        const itemURL = new URL(`${item.protocol}://${this.host}${item.pathname}`);
-        return itemURL.pathname === afterURL.pathname &&
-            itemURL.protocol === afterURL.protocol;
-      });
+  buildEndpoints() {
+    const endpoints = {};
 
-      if (afterIndex === -1) {
-        throw new Error(`${after} not found in test manifest`);
+    for (const [ident, test] of Object.entries(this.tests)) {
+      for (const scope of test.scope) {
+        let endpoint = '';
+        let httpsOnly = false;
+        switch (scope) {
+          case 'Window':
+            endpoint = '/api/interfaces';
+            break;
+          case 'Worker':
+          case 'DedicatedWorker':
+            endpoint = '/api/workerinterfaces';
+            break;
+          case 'ServiceWorker':
+            endpoint = '/api/serviceworkerinterfaces';
+            httpsOnly = true;
+            break;
+          case 'CSS':
+            endpoint = '/css/properties';
+            break;
+        }
+
+        if (endpoint) {
+          if (!(endpoint in endpoints)) {
+            endpoints[endpoint] = {
+              scope: scope,
+              httpsOnly: httpsOnly,
+              entries: []
+            };
+          }
+          if (!(ident in endpoints[endpoint].entries)) {
+            endpoints[endpoint].entries.push(ident);
+          }
+        }
       }
+    }
 
-      begin = afterIndex + 1;
-      if (limit) {
-        end = afterIndex + 1 + limit;
+    return endpoints;
+  }
+
+  listMainEndpoints(urlPrefix = '') {
+    return Object.keys(this.endpoints).map((item) => (
+      ['', `${urlPrefix}${item}`]
+    ));
+  }
+
+  listIndividual(urlPrefix = '') {
+    return Object.keys(this.tests).map((item) => (
+      [item, `${urlPrefix}/${item.replace(/\./g, '/')}`]
+    ));
+  }
+
+  listAllEndpoints(urlPrefix = '') {
+    return [
+      ...this.listMainEndpoints(urlPrefix),
+      ...this.listIndividual(urlPrefix)
+    ];
+  }
+
+  next(after) {
+    const afterURL = new URL(after);
+    if (!this.httpOnly && afterURL.protocol === 'http:') {
+      return `https://${this.host}${afterURL.pathname}`;
+    }
+    const endpoints = this.listMainEndpoints('/tests');
+    const index = endpoints.findIndex((item) => {
```

Comment:
What if `index` is -1 (before we add 1)? Isn't that an error case that we should do something about? As it is, will we end up running the same test in an endless cycle?

Reply:
Fixed down below!  If the `index` is 0 (`-1 + 1`), we return `null` now, to prevent an endless loop.